concurrency/StandardMatrixRequestGenerator.java

```diff
@@ -0,0 +1,123 @@
+
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.stream.Collectors;
+
+/**
+ * This is a class that replays predefined sequence of requests to generate requests.
+ * If it has already generated the last provided script, it would reuse the last request
+ * for all future requests of the same type until another request is registered.
+ * If a type of requests was never provided and is requested, it will throw an exception.
+ * @author chenshao
+ *
+ */
+@RequiredArgsConstructor
+public class StandardMatrixRequestGenerator implements MatrixRequestGenerator, Resettable
+{
+    private final Queue<GetMatrixRequest> getMatrixRequests = new ConcurrentLinkedQueue<>();
+    private final Queue<GetRevisionRequest> getRevisionRequests = new ConcurrentLinkedQueue<>();
+    private final Queue<PostRevisionRequest> postRevisionRequests = new ConcurrentLinkedQueue<>();
+    private final Queue<PutMatrixRequest> putMatrixRequests = new ConcurrentLinkedQueue<>();
+    private final Queue<DeleteMatrixRequest> deleteMatrixRequests = new ConcurrentLinkedQueue<>();
+    private final Queue<DiffRevisionRequest> diffRevisionRequests = new ConcurrentLinkedQueue<>();
+    
+    public StandardMatrixRequestGenerator withGetMatrixRequest(GetMatrixRequest... req)
+    {
+        getMatrixRequests.addAll(Arrays.stream(req).collect(Collectors.toList()));
+        return this;
+    }
+    
+    public StandardMatrixRequestGenerator withGetRevisionRequest(GetRevisionRequest... req)
+    {
+        getRevisionRequests.addAll(Arrays.stream(req).collect(Collectors.toList()));
+        return this;
+    }    
+    public StandardMatrixRequestGenerator withPutMatrixRequest(PutMatrixRequest... req)
+    {
+        putMatrixRequests.addAll(Arrays.stream(req).collect(Collectors.toList()));
+        return this;
+    }    
+    public StandardMatrixRequestGenerator withPostRevisionRequest(PostRevisionRequest... req)
+    {
+        postRevisionRequests.addAll(Arrays.stream(req).collect(Collectors.toList()));
+        return this;
+    }    
+    public StandardMatrixRequestGenerator withDiffRevisionRequest(DiffRevisionRequest... req)
+    {
+        diffRevisionRequests.addAll(Arrays.stream(req).collect(Collectors.toList()));
+        return this;
+    }    
+    public StandardMatrixRequestGenerator withDeleteMatrixRequest(DeleteMatrixRequest... req)
+    {
+        deleteMatrixRequests.addAll(Arrays.stream(req).collect(Collectors.toList()));
+        return this;
+    }
+    
+    @Override
+    public GetMatrixRequest getNextGetMatrixRequest(String matrixId)
+    {
+        GetMatrixRequest req = getMatrixRequests.peek();
+        if (Objects.isNull(req)) throw new RuntimeException("No GetMatrixRequest provided during construction.");
+        if (getMatrixRequests.size()==1) return req;
+        return getMatrixRequests.poll();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentLinkedQueue`, but your usage of `size()` and `poll()` may not be thread-safe at lines: **69 and 70**. If the `ConcurrentLinkedQueue` contains only **one element** and two threads perform this **same check at the same time** , the second thread&#39;s call to `poll()` will get `null`. This `null` may be used at line: **70**.

**Fix**  
 Consider calling `poll()`, checking instead of your current check if the returned object is `null`, and only then using that object.